### ISSUE TITLE
is there a way to prevent adding a background task twice ?

### ISSUE THREAD

**Issue by thomasleveil** [dutradda/apidaora#26]
I had a quick look at the code and it seems apidaora does not offer any mean of prevent a task with the same parameters to be queued up twice.

let say `POST /hello` from the example takes a minute to finish.

I'd looking for a way to have a first call to `curl -X POST -i localhost:8000/hello?name=Me` accept the task, but a second call to `curl -X POST -i localhost:8000/hello?name=Me` refusing the task if the first one is still in queue or running.

- does apidaora provide such a feature ?
- how could I hack apidaora to implement such a feature ?

**Comment by dutradda** [comment]
hello @thomasleveil !
Thank you for the question.

I tried to simulate what you said, and the api accept many POST requests from the same route with a large file.

I wrote the sample code (named `validate_post.py`):

```python
from typing import List

from apidaora import appdaora, route


@route.post('/hello')
async def hello_controller(body: List[str]) -> int:
    return len(body)


@route.get('/hello')
async def get_hello_controller() -> str:
    return 'Hello World!'


app = appdaora([hello_controller, get_hello_controller])
```

Then I started the server:

```bash
uvicorn validate_post:app
```

Then I generate a large file:

```bash
python -c "import json; open('large-array.json', 'w+').write(json.dumps(list(range(10000000))))"
du -sh large-array.json
```
```
85M    large-array.json
```

Then I open 4 terminals, and execute the following call:
```bash
curl -X POST -i localhost:8000/hello -T large-array.json
```

all the 4 calls was accept by the server.

And I open a fifth terminal and execute the following call:
```bash
curl -i localhost:8000/hello
```

That was accepted too.


Can you share yours tests to I try to found the problem?

**Comment by dutradda** [comment]
Hello @thomasleveil !

If understood your question right, apidaora does not have feature to prevent the same request to be called more then once.

But I think it is possible to be implement.
You know others frameworks that implement this feature?

**Comment by dutradda** [comment]
@thomasleveil 
I think we can do it using the [middlewares](https://dutradda.github.io/apidaora/using-middlewares/) module, I will try to do it and I will return to you

**Comment by thomasleveil** [comment]
Hi @dutradda 

It is quite difficult to find such a feature in this realm. Here are a few examples :
- the [`unique`](https://github.com/closeio/tasktiger#task-options) option of TaskTiger
- the RabbitMQ [message-deduplication](https://github.com/noxdafox/rabbitmq-message-deduplication) plugin
- [this](http://docs.celeryproject.org/en/latest/tutorials/task-cookbook.html#ensuring-a-task-is-only-executed-one-at-a-time) Celery recipe

**Comment by dutradda** [comment]
@thomasleveil 
I have implemented a simple middleware that prevents the request to be called until the body was processed.

Here is an example how to use it: https://dutradda.github.io/apidaora/using-prevent-request-middleware/

The code is here: https://github.com/dutradda/apidaora/blob/master/apidaora/middlewares.py#L117

**Comment by dutradda** [comment]
> Also, I really like how you implemented this backgroundTask system and I wonder if it could be extracted into its own python module so that it could be used with any ASGI server.

This feature is coupled with apidaora code. I don't know how difficult is to apart it. I will analyse it.

I used the background task feature on the chunli app: https://dutradda.github.io/chunli/

**Comment by dutradda** [comment]
@thomasleveil 
I can implement the blocking request on the background tasks feature so.
It is what you want, right?

**Comment by thomasleveil** [comment]
> I can implement the blocking request on the background tasks feature so.

yes, I want the endpoint to refuse queue a second task with the same parameter if the first task is not processed (the task, not the creation request)

**Comment by dutradda** [comment]
hello @thomasleveil 

I have implemented the feature you request! 🎉 

Take a look on documentation and tell me what you think: https://dutradda.github.io/apidaora/using-redis-background-task-lock-args-controller/

**Comment by thomasleveil** [comment]
Thank you for considering this feature :)

I gave it a try, and I'm confused with task_id being the same when running two tasks with the same parameters. For both tasks, the `task_key` should indeed be the same in order to perform the deduplication ; but `task_id` should be unique per task.

You can clone [this](https://gist.github.com/thomasleveil/f7e0e24e00415e2295d020fa88d191a8) gist to run my tests

**Comment by dutradda** [comment]
> For both tasks, the task_key should indeed be the same in order to perform the deduplication ; but task_id should be unique per task.

You are right @thomasleveil , I will change it to be unique too!
Thanks for your feedback!

**Comment by thomasleveil** [comment]
I'm also uneasy with the `HTTP 400` status code returned when creating a task for the 2nd time. `4xx` statuses are for _client errors_, but there is nothing the client did wrong. 

searching the web for eventual standards, I was unable to find anything close to a standard for status code related to async tasks, but I stumbled upon different sources which suggest to use the [`HTTP 303 See Other`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/303) status code.

What I like about the `HTTP 303` response is that it must include the `Location` header having for value the url of the already existing task. Apidaora could provide the url for `/hello-single?task_id=<id of 1st task>`

_On a side note, some say it is more correct (in regard to REST) to respond with `HTTP 202 Accepted` rather than `HTTP 201 Created` for endpoints creating a async task_

**Comment by dutradda** [comment]
@thomasleveil
I have implemented the unique task-id for lock args: https://dutradda.github.io/apidaora/background-task-controller/lock-args-redis/

I will implement the 202 status code, is more correct than 201.

about the 303, I will think about it, and search for others status too.

**Comment by thomasleveil** [comment]
Here are a few links regarding asynchronous REST : 

- https://www.adayinthelifeof.nl/2011/06/02/asynchronous-operations-in-rest/
- https://farazdagi.com/2014/rest-and-long-running-jobs/
- https://stackoverflow.com/questions/36421707/http-status-code-for-async-tasks/36422103

**Comment by thomasleveil** [comment]
I don't know if it is intentional but the 200 response when getting the status of a **finished** task does not include the `args_signature` property

**Comment by dutradda** [comment]
@thomasleveil 
I add the 202 and 303 status code

I fix the finished task too, adding the args_signature to response.

https://dutradda.github.io/apidaora/background-task-controller/lock-args-redis/

**Comment by thomasleveil** [comment]
Excellent :tada:  

I took the time to test it and found no issue with the locking behavior. Thank you again for your quick reaction to this issue :+1:

**Comment by thomasleveil** [comment]
actually, I found an issue when adding two tasks from different modules and those task associated functions have the same name : 

```python
# module1.py
import time

from apidaora import route


@route.background('/hello-single-1', lock_args=True)
def hello_task(name: str) -> str:
    time.sleep(1)
    return f'Hello {name} from module 1!'
```

```python
# module2.py
import time

from apidaora import route


@route.background('/hello-single-2', lock_args=True)
def hello_task(name: str) -> str:
    time.sleep(1)
    return f'Hello {name} from module 2!'
```

```python
# myapp.py
from apidaora import appdaora
import module1
import module2

app = appdaora([
    module1.hello_task,
    module2.hello_task
])
```

```bash
uvicorn myapp:app &
curl -X POST -i localhost:8000/hello-single-1?name=Me  # 202
curl -X POST -i localhost:8000/hello-single-2?name=Me  # 303
```

----

The fix might just be as easy as to replace `controller.__name__` with `controller.__qualname__`.

Or maybe use the request path instead of the controller function name to build the task key ?

_You can [clone this gist](https://gist.github.com/thomasleveil/f7e0e24e00415e2295d020fa88d191a8) to run my tests_
